### PR TITLE
Power Service Implementation

### LINKED ISSUE #2334: Implement Power Service
### Description of the task

Respond to auto kick/chip commands by sending the appropriate UART message. 

### Acceptance criteria

- [ ] Implement PowerService from (software/jetson_nano/services/service.h)
- [ ] TODO

### PR DESCRIPTION
**Pull request by LiCody**
### Description

- Power service takes in a power command that is sent over uart and reads power statuses over the same uart on a separate thread.
- Modify uart_framing to not use boost::optional as external libraries are a pain with platformio

### Testing Done

Tested manually with jetson and esp32 with test programs to check that commands are sent correctly and statuses are read correctly

### Resolved Issues
resolves #2334

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/jetson_nano/services/power.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     * Service that interacts with the power board.
+     * Opens all the required ports and maintains them until destroyed.
+     */
+    PowerService();
+    ~PowerService();
+    void start() override;
+    void stop() override;
+    /**
+     * When the power service is polled it sends the given power and
+     * returns the latest power status
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/services/power.h]: nit: given power -> given power command

File: src/software/jetson_nano/services/power.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * @param command The
+     * @return
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/services/power.h]: nit: unfininished comment

File: src/software/jetson_nano/services/power.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+std::unique_ptr<PowerStatus> PowerService::poll(const PowerCommand& command)
+{
+    auto frame        = createUartMessageFrame(command);
+    auto write_buffer = frame.marshallUartPacket();
+
+    int i = 0;
+    while (i < MAX_WRITE_ATTEMPTS)
+    {
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/services/power.cpp]: curious why this is required?

File: src/software/jetson_nano/services/power.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        auto uart_frame = UartMessageFrame<PowerStatus>();
+        if (unmarshalUartPacket<PowerStatus>(power_status_msg, uart_frame))
+        {
+            LOG(DEBUG) << "Command status read successfully";
+            status = uart_frame.message;
+        }
+        else
+        {
+            LOG(WARNING) << "Unmarshal failed";
+        }
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/services/power.cpp]: can this be in-line in the poll function and not in a separate thread?

File: src/software/jetson_nano/services/power.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    std::atomic_bool in_destructor = false;
+    std::atomic<PowerStatus> status;
+
+    // Constants
+    const size_t READ_BUFFER_SIZE;
+    static constexpr unsigned int BAUD_RATE                 = 115200;
+    const std::string DEVICE_SERIAL_PORT                    = "/dev/ttyTHS1";
+    static constexpr int MAX_WRITE_ATTEMPTS                 = 3;
+    static constexpr unsigned int INTERVAL_BETWEEN_READS_MS = 5;
+    static constexpr unsigned int STARTUP_TIME_MS           = 0;
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/services/power.h]: you don't need the `static constexpr` here just `const unsigned` should be enough

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
:zap: :electric_plug:

File: src/software/jetson_nano/services/power.cpp (hunk 4 above)
Comment (LiCody) [reply src/software/jetson_nano/services/power.cpp]: Thinking about it again it probably can be. I was assuming thunderloop would only send the power status when it is sent a command but its probably easier on both ends to have thunderloop continuously send power statuses and move reading into the poll fn.

File: src/software/jetson_nano/services/power.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <boost/asio.hpp>
+
+#include "software/jetson_nano/services/service.h"
+#include "software/logger/logger.h"
+#include "software/uart/boost_uart_communication.h"
+
+// Move different file with powerloop
```
Comment (amrm3lm) [inline src/software/jetson_nano/services/power.h]: not sure what this means? is it a todo?

**Review by amrm3lm — COMMENTED** [review]
lgtm just one clarification point

File: src/software/jetson_nano/services/power.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            LOG(WARNING) << "Writing power command failed.";
+        }
+
+        // Read power status
+        uart->flushSerialPort(uart->flush_receive);
+        power_status_msg = uart->serialRead(READ_BUFFER_SIZE);
+    }
+    catch (std::exception& e)
+    {
+        LOG(FATAL) << "ESP32 has disconnected. Power service has crashed" << e.what();
```
Comment (akhilveeraghanta) [inline src/software/jetson_nano/services/power.cpp:42]: we should technically be able to run without kick/chip, but its not clear what we would do here rn

eventually this should be an error code, but we can revisit it

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm! lets do that protobuf thing in another PR

**Review by akhilveeraghanta — DISMISSED** [review]

File: src/software/jetson_nano/services/power.h [hunk 8]
```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <boost/asio.hpp>
+
+#include "software/jetson_nano/services/service.h"
+#include "software/logger/logger.h"
+#include "software/uart/boost_uart_communication.h"
+
+// TODO: change to protobuf/move to different file with addition of thunderloop
```
Comment (jonathanlew) [inline src/software/jetson_nano/services/power.h]: please tag this TODO with an issue if it's still relevant

**Review by jonathanlew — COMMENTED** [review]
as a sanity check, can you check that things still work with https://github.com/UBC-Thunderbots/Software/pull/2535

**Review by jonathanlew — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**